### PR TITLE
python3Packages.flash-attn: init at 2.6.3

### PR DESCRIPTION
**Pull request by cfhammill**
## Description of changes

Adds the flash-attn package. Similar to https://github.com/NixOS/nixpkgs/pull/341195, only cuda has been tested (84 F, 300P S, 200K S; yes it has 500 thousand tests). The 84 failures are numerical and likely just numerical instability. (This was tested with the triton fix from https://github.com/NixOS/nixpkgs/pull/344259).

This can be used as an optional dependency for vllm and likely other packages as well.

cc @SomeoneSerge @samuela @natsukium 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/flash-attn/default.nix [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    repo = "flash-attention";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-9AmmoUlyh4x0CuasZeXML7N3TETmG0vhHMm8cirsblk=";
+  };
+
+  build-system = [
+    ninja
+    setuptools
+    torch
+    psutil
```
Comment (samuela) [inline pkgs/development/python-modules/flash-attn/default.nix:44]: nit: alphabetize

File: pkgs/development/python-modules/flash-attn/default.nix [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  ];
+
+  nativeBuildInputs = [ which git ];
```
Comment (samuela) [inline pkgs/development/python-modules/flash-attn/default.nix:47]: nit: alphabetize

File: pkgs/development/python-modules/flash-attn/default.nix [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    lib.optionals cudaSupport (
+      with cudaPackages;
+      [
+        cuda_cudart # cuda_runtime.h, -lcudart
+        cuda_cccl
+        libcusparse # cusparse.h
+        libcusolver # cusolverDn.h
+        cuda_nvcc
+        libcublas
+        cutlass
```
Comment (samuela) [inline pkgs/development/python-modules/flash-attn/default.nix:59]: nit: alphabetize

File: pkgs/development/python-modules/flash-attn/default.nix [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  psutil
+}:
+
+let
+  cutlass = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "cutlass";
+    rev = "refs/tags/v3.5.0";
+    sha256 = "sha256-D/s7eYsa5l/mfx73tE4mnFcTQdYqGmXa9d9TCryw4e4=";
+  };
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/flash-attn/default.nix:25]:
```
❯ ag cutlass
pkgs/development/libraries/onnxruntime/update-re2.patch
43: cutlass;https://github.com/NVIDIA/cutlass/archive/refs/tags/v3.1.0.zip;757f90a795034a89d4f48a79d1f009f7a04c8dee

pkgs/development/libraries/onnxruntime/default.nix
83:   cutlass = fetchFromGitHub {
85:    repo = "cutlass";
205:    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_CUTLASS" "${cutlass}")

pkgs/development/python-modules/vllm/default.nix
51:  cutlass = fetchFromGitHub {
53:    repo = "cutlass";
156:  cmakeFlags = [ (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_CUTLASS" "${lib.getDev cutlass}") ];
```

It's maybe time we moved it to `cudaPackages`...

File: pkgs/development/python-modules/flash-attn/default.nix [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  fetchFromGitHub,
+  einops,
+  ninja,
+  setuptools,
+  torch,
+  transformers,
+  triton,
+  cudaPackages,
+  config,
+  cudaSupport ? config.cudaSupport,
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/flash-attn/default.nix:13]: If the package is cuda-only (no rocm and no cpu for example) we might as well ignore these flags (except for `cudaCapabilities`)

File: pkgs/development/python-modules/flash-attn/default.nix [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  # pytest tests not enabled due to nvidia GPU dependency
+  pythonImportsCheck = [ "flash_attn" ];
+
+  meta = with lib; {
+    description = "Fast and Memory-Efficient Exact Attention with IO-Awareness";
+    homepage = "https://github.com/Dao-AILab/flash-attention";
+    license = licenses.bsd3;
+    maintainers = with maintainers; [ cfhammill ];
+    # The package requires CUDA or ROCm, the ROCm build hasn't
+    # been completed or tested, so broken if not using cuda.
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/flash-attn/default.nix:83]: Ah I see the message now

File: pkgs/development/python-modules/flash-attn/default.nix [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  build-system = [
+    ninja
+    setuptools
+    torch
+    psutil
+  ];
+
+  nativeBuildInputs = [ which git ];
+
+  buildInputs = (
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/flash-attn/default.nix:49]: redundant parens btw

**Comment by SomeoneSerge** [comment]
Ofborg's nixfmt complaints would need to be addressed, but otherwise I think this is mergeable. But @cfhammill could you provide some input on what is the benefit of adding flash_attn to Nixpkgs, are there any reverse dependencies that need it or just a particular user-base? There's already an implementation of flash attention in pytorch, so there must be something specific about this upstream? Thanks

**Comment by SomeoneSerge** [comment]
@cfhammill hi, sorry to bump, are you interested in getting this merged?

**Comment by cfhammill** [comment]
@SomeoneSerge I don't think I can justify adding it, flash-attn2 was added to transformers, and the use-case I had uses their own fork.

**Comment by SomeoneSerge** [comment]
I think I recently saw some downstream packages using this (iirc: chroma, unsloth). Also I didn't mean to leave the impression that there's some formal requirement of "justification", just reflecting on where we're going

**Pull request closed without merging**